Content.Shared/Actions/SharedActionsSystem.cs

```diff
@@ -69,8 +69,47 @@ public override void Initialize()
         SubscribeAllEvent<RequestPerformActionEvent>(OnActionRequest);
     }
 
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        // All of this is extremely expensive, please make sure it's not active conssistently
+
+        var worldActionQuery = EntityQueryEnumerator<WorldTargetActionComponent>();
+        while (worldActionQuery.MoveNext(out var uid, out var action))
+        {
+            if (IsCooldownActive(action) || !ShouldResetCharges(uid, action))
+                return;
+
+            UpdateAction(uid, action);
+            Dirty(uid, action);
+        }
+
+        var instantActionQuery = EntityQueryEnumerator<InstantActionComponent>();
+        while (instantActionQuery.MoveNext(out var uid, out var action))
+        {
+            if (IsCooldownActive(action) || !ShouldResetCharges(uid, action))
+                return;
+
+            UpdateAction(uid, action);
+            Dirty(uid, action);
+        }
+
+        var entityActionQuery = EntityQueryEnumerator<EntityTargetActionComponent>();
+        while (entityActionQuery.MoveNext(out var uid, out var action))
+        {
+            if (IsCooldownActive(action) || !ShouldResetCharges(uid, action))
+                return;
+
+            UpdateAction(uid, action);
+            Dirty(uid, action);
+        }
+    }
+
```

Comment:
Few things:
1. You're using returns instead of continues which will break a heap of things.
2. Server doesn't actually care need to care about this. Server only cares when someone requests an action. You can derive whether something has charges / cooldown easily without spamming constant network states and flooding the network.
3. Visual updates should be done in FrameUpdate, not Update.